### PR TITLE
fix(types): improve `strict` function type

### PR DESCRIPTION
**Pull request by Mister-Hope**

### CODE REVIEW

File: types/katex.d.ts [hunk 1]
```diff
@@ -20,6 +20,43 @@ export type TrustContext =
...
+  errorCode:
+    | "unknownSymbol"
+    | "unicodeTextInMathMode"
+    | "mathVsTextUnits"
+    | "commentAtEnd"
+    | "htmlExtension"
+    | "newLineInDisplayMode",
+  errorMsg: string,
+  token: KatexToken,
+) => boolean | "error" | "warn" | "ignore" | undefined | null | void;
```
Comment (Mister-Hope) [inline types/katex.d.ts:57]: Actually I think `boolean | "error" | "warn" | "ignore" | void` is enough

File: types/katex.d.ts [hunk 2]
```diff
@@ -20,6 +20,43 @@ export type TrustContext =
...
+  catcodes: KatexCatcodes;
+}
+
+interface KatexSourceLocation {
+  start: number;
+  end: number;
+  lexer: KatexLexerInterFace;
+}
+
+export interface KatexToken {
```
Comment (Mister-Hope) [inline types/katex.d.ts:39]: BTW, to safely fit all users even they are patching katex for they own usage, we can let katexToken extends `Record<string, unknown>`, so other properties hacked by users can be added without causing any type issues.

Requesting for a feedback here.

File: types/katex.d.ts [hunk 3]
```diff
@@ -20,6 +20,43 @@ export type TrustContext =
     | { command: "\\htmlStyle", style: string }
     | { command: "\\htmlData", attributes: Record<string, string> }
 
+
+type KatexCatcodes = Record<string, number>;
```
Comment (edemaine) [inline types/katex.d.ts:24]: I don't think you need to prefix all the types and interfaces with `Katex`. And probably they should all be `export`ed too.

File: types/katex.d.ts [hunk 4]
```diff
@@ -20,6 +20,43 @@ export type TrustContext =
...
+
+interface KatexLexerInterFace {
```
Comment (edemaine) [inline types/katex.d.ts:26]: I also don't think you need `Interface` in the name of this `Interface`. Matching the internal `Lexer` seems fine.

File: types/katex.d.ts [hunk 5]
```diff
@@ -20,6 +20,43 @@ export type TrustContext =
...
+
+interface KatexSourceLocation {
+  start: number;
+  end: number;
+  lexer: KatexLexerInterFace;
+}
+
+export interface KatexToken {
+  text: string;
+  loc: KatexSourceLocation;
```
Comment (edemaine) [inline types/katex.d.ts:41]: According to `Token.js`, `loc` should have `| undefined` as well.

File: types/katex.d.ts [hunk 6]
```diff
@@ -20,6 +20,43 @@ export type TrustContext =
...
+  noexpand: boolean | undefined;
+  treatAsRelax: boolean | undefined;
```
Comment (edemaine) [inline types/katex.d.ts:43]: Actually, in TypeScript notation, I think these should be
```ts
  noexpand?: boolean;
  treatAsRelax?: boolean;
```

File: types/katex.d.ts (hunk 1 above)
Comment (edemaine) [reply types/katex.d.ts:57]: I don't think `| void` means what you think it means... I definitely wouldn't write that.

I agree that the docs don't allow `undefined | null` but the code explicitly says that it does. I could go either way on including them; probably `boolean` is enough. Or adding just `undefined` is maybe helpful.

File: types/katex.d.ts [hunk 7]
```diff
@@ -20,6 +20,43 @@ export type TrustContext =
...
+}
+
+
+export type KatexLoggerFunction = (
```
Comment (edemaine) [inline types/katex.d.ts:47]: I think this should be called `StrictFunction`

File: types/katex.d.ts (hunk 2 above)
Comment (edemaine) [reply types/katex.d.ts:39]: I don't think that's necessary; the user can always [merge interfaces](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#merging-interfaces) to extend anything we provide. (You'd need to wrap it in `declare module "katex" { ... }` I believe.)

**Review by edemaine — CHANGES_REQUESTED** [review]
Looks like a good start! Here are some suggested changes.

**Comment by Mister-Hope** [comment]
All done

**Review by edemaine — APPROVED** [review]
Thanks!

**Pull request merged**